About us

kriststoja opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Adding an about me page for Artemis

### Description
Created an about me page that is reachable from the footer and without needing authentication.
Created an json file under resources which will be read from the about me component when it is opened, also adding user photos in the resources page
Created the object structure for reading the json file

### Steps for Testing

1. Open Artemis
2. On the footer (left) press About Me

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![image](https://user-images.githubusercontent.com/75738365/101647555-26986300-3a39-11eb-826c-92d452d92b3f.png)
![image](https://user-images.githubusercontent.com/75738365/101647646-43cd3180-3a39-11eb-854a-bc1d54d1825e.png)
![image](https://user-images.githubusercontent.com/75738365/101647704-547da780-3a39-11eb-8c8f-48415d05a34b.png)

FrankeLukas commented on src/main/webapp/app/core/about-us/about-us-routing.module.ts:
```diff
@@ -0,0 +1,26 @@
+
+const routes: Routes = [
+    {
+        path: 'about',
+        children: [
+            {
+                path: '',
+                pathMatch: 'full',
```
This looks a bit weird. Why aren't we defining component and data directly? Along the lines of just
```
path: 'about',
component: AboutUsComponent,
data: {
    authorities: [],
    pageTitle: 'aboutUs',
},
```

kriststoja replied:
Yes you are right, its weird. I started it with the idea I would need path children and forgot to adjust it later when it was not needed. Ill change it now

stefanwaldhauser reviewed (changes requested):
Is it possible to also write a quick client test for this :) ? I can also quickly do it myself if you want 👍🏻 

I get the following error on ts1

![image](https://user-images.githubusercontent.com/29718932/101686562-84dd3a00-3a69-11eb-8640-57ee3f37e664.png)


https://sentry.io/organizations/ls1intum/issues/1929849936/?environment=test&project=1440029&query=&statsPeriod=14d

stefanwaldhauser commented on src/main/webapp/app/core/about-us/about-us.component.ts:
```diff
@@ -0,0 +1,61 @@
+     * On init get the mail data needed for the contact
+     */
+    ngOnInit(): void {
+        this.staticContentService.getStaticJsonFromArtemisServer('/about-us.json').subscribe((data) => {
```
So it looks like `private readonly staticContentUrl = SERVER_API_URL + 'public/content/';` therefore this will create `public/content//about-us.json` 
```suggestion
        this.staticContentService.getStaticJsonFromArtemisServer('about-us.json').subscribe((data) => {
```

kriststoja replied:
Fixed it

stefanwaldhauser commented:
![image](https://user-images.githubusercontent.com/29718932/101687478-d934e980-3a6a-11eb-9f13-64bd7b4311bc.png)
Is the email correct here? Should it not be `artemis@in.tum.de`?

stefanwaldhauser commented on src/main/webapp/app/core/about-us/about-us.component.ts:
```diff
@@ -0,0 +1,61 @@
+            .pipe(
+                filter(Boolean),
+                tap((info: ProfileInfo) => {
+                    this.contact = info.contact;
```
Just out of interest. Why are you doing this with a `tap` operator here and not in the subscribe part?

kriststoja replied:
As I wrote below, this part was only imported from the footer component since some parts of it were moved inside about us page. I could give an assumption why it was done like this, but wouldnt be the answer of the one who did it :)

As for the email address it would be the same answer. I gave the benefit of the doubt it is correct, if not it has been for a long time wrong and I could fix it with this case

stefanwaldhauser commented on src/main/webapp/app/core/about-us/about-us.component.ts:
```diff
@@ -0,0 +1,61 @@
+        this.profileService
+            .getProfileInfo()
+            .pipe(
+                filter(Boolean),
```
Just trying to learn more about rxjs? Why are you doing this here? Is it possible that getProfileInfo sends undefined?

kriststoja replied:
Small explanation. This implementation here, together with the method set contact() are not done by me. They were implemented on footer component, but since Report a bug, Request a feature and Contact us are now inside the about page, I had to take this logic with in this component. I didnt overanalyze this part and gave the benefit of the doubt that when they were implemented back then, they have already been checked. That would be an answer also to the previous 2 comments.

As for rxjs question: getProfileInfo() here is a BehaviourSubject, an Observable used to stream data. We can subscribe to it and it always delivers immediately the latest data when data changes. The good of it is it will always send a valid value, even if no data has been emitted, thats why the risk of an error because of the undefined is eleminated. Thats why BehaviourSubjects are normaly initiated with undefined or null values (or empty values in general). It is actually also very good when you want to render the html parts only when data is there, not before, and can use BehaviourSubjetcs with async pipe to make sure of it.

airelawaleria reviewed (approved):
Tested on ts1. Looks good! The only thing is that there are some pictures missing, but I guess it's not an issue.

marvinchimeka reviewed (approved):
tested on ts1 with @airelawaleria. Some pictures are missing. beside that everything works as intended.

kriststoja commented:
> tested on ts1 with @airelawaleria. Some pictures are missing. beside that everything works as intended.

I got the names and pictures from github list. Some of the persons dont have one there, thats why I used placeholders. We can add them later if someone provides them to me :)

kriststoja commented:
> Tested on ts1. Looks good! The only thing is that there are some pictures missing, but I guess it's not an issue.

I didnt have the data for the ones missing, since I got them from the list in github. We can add them on later if I get the data :)

JohannesWeiss reviewed (dismissed):
One image failed to load, but otherwise everything looks good.
![image](https://user-images.githubusercontent.com/37304775/101912597-4313d700-3bc2-11eb-92fc-3daf9bfa7d3f.png)

kriststoja commented:
> One image failed to load, but otherwise everything looks good.
> ![image](https://user-images.githubusercontent.com/37304775/101912597-4313d700-3bc2-11eb-92fc-3daf9bfa7d3f.png)

kriststoja commented:
> Is it possible to also write a quick client test for this :) ? I can also quickly do it myself if you want 👍🏻
> 
> I get the following error on ts1
> 
> ![image](https://user-images.githubusercontent.com/29718932/101686562-84dd3a00-3a69-11eb-8640-57ee3f37e664.png)
> 
> https://sentry.io/organizations/ls1intum/issues/1929849936/?environment=test&project=1440029&query=&statsPeriod=14d

I have added a test class

airelawaleria reviewed (dismissed):
Re-tested the last changes. Looks good!

simonzilker reviewed (dismissed):
Looks good!

stefanwaldhauser reviewed (changes requested):
GJ I just have two things to note if you want to fix them:
- ![image](https://user-images.githubusercontent.com/29718932/101928138-83c91b80-3bd5-11eb-8134-2ad2414fe7a8.png) Maybe you could use a placeholder for a human here if the image does not exist(like a silhoute of a human) and center the image
- The Page is broken on mobile screen sizes:
![image](https://user-images.githubusercontent.com/29718932/101928449-e1f5fe80-3bd5-11eb-85de-ae610524d6a1.png)

Please make it responsive :)

valeriebucher reviewed (dismissed):
Tested on ts1, looks nice!

jpbernius commented on src/main/webapp/app/core/about-us/about-us-routing.module.ts:
```diff
@@ -0,0 +1,20 @@
+    imports: [RouterModule.forChild(routes)],
+    exports: [RouterModule],
+})
+export class AboutUsRoutingModule {}
```
Please lazy-load this module OR make this part of another module we already have.

kriststoja replied:
Done. With lazy load

kriststoja replied:
Responsive layout also

jpbernius commented on src/main/webapp/app/core/about-us/artemis-about-us.module.ts:
```diff
@@ -0,0 +1,11 @@
+    declarations: [AboutUsComponent],
+    imports: [AboutUsRoutingModule, TranslateModule, CommonModule],
+})
+export class ArtemisAboutUsModule {}
```
Why do we need two modules? One should be enough? (And only if we end up lazy loading, otherwise this should not be needed at all for a single component.

kriststoja replied:
Fixed

jpbernius commented on src/test/javascript/spec/component/about-us/about-us.component.spec.ts:
```diff
@@ -0,0 +1,59 @@
+            .compileComponents()
+            .then(() => {
+                fixture = TestBed.createComponent(AboutUsComponent);
+                component = fixture.componentInstance;
```
```suggestion
```

kriststoja replied:
removed

jpbernius commented on src/test/javascript/spec/component/about-us/about-us.component.spec.ts:
```diff
@@ -0,0 +1,59 @@
+const expect = chai.expect;
+
+describe('AboutUsComponent', () => {
+    let component: AboutUsComponent;
```
```suggestion
```

jpbernius commented on src/main/webapp/app/core/about-us/about-us.component.html:
```diff
@@ -0,0 +1,58 @@
+<div class="mt-5" *ngIf="data && data.contributors">
+    <h3 class="mb-3 font-weight-bold">{{ 'aboutUsOverview.contributors' | translate }}</h3>
+    <div class="row">
+        <div class="col-sm-1 grid-margin" *ngFor="let contributor of data.contributors">
```
Is this ordered in any particular order? What about a random order on each pageload?

kriststoja replied:
The json file was build based on the sorting of github contributors, I think there its ordered based on amount of commits. Of course I can list the contributors randomly if you prefer so

jpbernius reviewed (changes requested):
From the screenshots, this is looking nice. I second @stefanwaldhauser for a mobile-friendly responsive layout.
Also, some small suggestions from the code.
In general, this is already looking good. Keep it up!

tiffca reviewed (dismissed):
Tested on test server 1 successfully, changes look good!

kriststoja commented:
> GJ I just have two things to note if you want to fix them:
> 
> * ![image](https://user-images.githubusercontent.com/29718932/101928138-83c91b80-3bd5-11eb-8134-2ad2414fe7a8.png) Maybe you could use a placeholder for a human here if the image does not exist(like a silhoute of a human) and center the image
> * The Page is broken on mobile screen sizes:
>   ![image](https://user-images.githubusercontent.com/29718932/101928449-e1f5fe80-3bd5-11eb-85de-ae610524d6a1.png)
> 
> Please make it responsive :)

Fixed it.

I also added person placeholders for missing photos. At first didn't want to just get some random photo from internet, but I found something. Actually we must have photos for everyone, maybe we can fill it later.

JohannesWeiss reviewed (dismissed):
Tested on ts3 with responsive simulator, works great now on all kinds of relevant devices. Placeholder images look better now as well. Re-Approve.

airelawaleria reviewed (dismissed):
Looks good now :)

simonzilker reviewed (dismissed):
Looks good! Tested also the responsive design! Re-approve

EneaGore reviewed (approved):
Looks really good rn.

stefanwaldhauser reviewed (approved):
Looks good to me now :) Gj

anditurdiu reviewed (dismissed):
Code looks good, tested locally also

philgabr reviewed (approved):
Looks nice and works as intended (also the responsive design)!

ivanchimeno reviewed (dismissed):
Looks good!

FuchsDominik reviewed (dismissed):
Code as well as (responsive) page looks good 👍

marvinchimeka reviewed (approved):
Looks very nice and clean. Approve.

JannisBittlmayer reviewed (approved):
Looks great and works an intended.

FlorianStang reviewed (dismissed):
Tested on TS2, everything worked. Nice feature! 👍

stefanwaldhauser reviewed (approved):
Reapprove

airelawaleria reviewed (approved):
Re-approve

JohannesWeiss reviewed (approved):
Re-Approve, still works as expected.

marvinchimeka reviewed (approved):
re-approve

simonzilker reviewed (approved):
re-approve!

jpbernius reviewed (approved):
LGTM :shipit:
